File: Practice/AllDomains/CoreCS/DataStructures/Stacks/MaximumElement.js
/*	author:@shivkrthakur  */
var index = 0;

function readLine(input){
    return input[index++];    
}

function processData(input) {
    //Enter your code here
    var n = parseInt(readLine(input));
    var stack = [], maxStack = [];
    for(var i = 0; i < n; i++){
      var arr = readLine(input).split(' ').map(Number);
      var x = arr[0];
      if(x === 1) {
        var val = arr[1];
        stack.push(val);
        if(maxStack.length > 0){
          if(maxStack[maxStack.length - 1] <= arr[1]) maxStack.push(val);
        }
        else maxStack.push(val);
      }
      else if(x == 2){
        var val = stack.pop();
        if(maxStack[maxStack.length - 1] == val) maxStack.pop();
      }
      else {
        console.log(maxStack[maxStack.length - 1]);
      }
    }
} 

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input.split('\n'));
});

File: Practice/AllDomains/CoreCS/DataStructures/Stacks/MaximumElement.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
class Solution {
    static void Main(String[] args) {
        int n = Convert.ToInt32(Console.ReadLine());
        Stack<int> stack = new Stack<int>();
        Stack<int> maxStack = new Stack<int>();
        maxStack.Push(0);
        for(int i = 0; i < n; i++){
            string input = Console.ReadLine();
            int query = Convert.ToInt32(input[0].ToString());

            if(query == 1)
            {
                stack.Push(Convert.ToInt32(input.Split()[1]));
                if(stack.Peek() >= maxStack.Peek()) maxStack.Push(stack.Peek());
            }
            else if(query == 2)
            {
                if(maxStack.Peek() == stack.Peek()) maxStack.Pop();
                stack.Pop();
            }
            else Console.WriteLine(maxStack.Peek());
        }
    }
}
